### ISSUE TITLE
Add possibility to modify the powergrid python side: loads, generators, storage units and shunts

### ISSUE THREAD

**Issue by BDonnot** [powsybl/pypowsybl#79]
## **Do you want to request a *feature* or report a *bug*?**

This is a feature request.

## **What is the current behavior?**

Today, i did not see any way to change the grid from python side, beside the opening / closing of switches detailed in the example of the readme.

## **What is the motivation / use case for changing the behavior?**
This would allow for a great interoperability with grid2op (https://github.com/rte-france/Grid2Op) where the loads / generators are changed a lot: we represent the "time" and at each "time step" the productions / loads / shunts / storage units are modified.

## **What is the expected behavior?**

For example, i would like to be able to:

- change a load (active or reactive consumption )
- change a generator (active production or voltage setpoint)
- change a storage unit (active production / consumption)
- change a shunt (active / reactive)

This would give, for unitary changes, something like
```python
import pypowsybl as pp
net = pp.load(...)
net.change_load(load_name_or_id, new_p, new_q)
net.change_gen(gen_name_or_id, new_p, new_v)
net.change_storage(storage_name_or_id, new_p)
net.change_shunt(storage_name_or_id, new_p)
```

And for performance reasons, it would also be interesting to do it in a "vectorized" fashion:
```python
import pypowsybl as pp
net = pp.load(...)
net.change_loads(vector_of_load_name_or_id, vector_of_new_p, vector_of_new_q)
# same for `change_gens` and `change_storages`
```

## **Please tell us about your environment:**
  - PowSyBl Version: `all`
  - OS Version: `all`
Ideally, all version would be concerned.

**Comment by sylvlecl** [comment]
Not closing yet the issue, but there has been evolutions in that direction: we'll now be able to update the network using pandas dataframes, see for example this extract from unit tests:
```python
    def test_update_generators_data_frame(self):
        n = pp.network.create_eurostag_tutorial_example1_network()
        ...
        df2 = pd.DataFrame(data=[[608.0, 302.0, 25.0, False]], columns=['target_p','target_q','target_v','voltage_regulator_on'], index=['GEN'])
        n.update_generators_with_data_frame(df2)
        df3 = n.create_generators_data_frame()
        self.assertEqual(608, df3['target_p']['GEN'])
        self.assertEqual(302.0, df3['target_q']['GEN'])
        self.assertEqual(25.0, df3['target_v']['GEN'])
        self.assertFalse(df3['voltage_regulator_on']['GEN'])
```

Does it look OK to you?

**Comment by BDonnot** [comment]
I did not had the time to try it yet (i probably need to install pypowsybl from source again and did not do that yet) but it looks good.

One thing that might be difficult to do: are we forced to provide all the data (in this case both "target_p", "target_q" and "target_v") in the dataframe ? Is there a way to have a "default" value (for example filling with `np.Nan`) when we don't want to modify all the property of a generator (say i want to only modify the voltage setpoint and not the active production for example)

**Comment by sylvlecl** [comment]
You may rebuild from sources, or wait for the release which should happen shortly :) 

All columns of the dataframe are optional, so you may define only the fields you need to change. However, it does not allow, indeed, to update different fields for different equipments in the same call (for example `target_v` on generator G1 and `target_q` on generator G2).

**Comment by geofjamg** [comment]
@BDonnot right, we should support the nan to allow partial update in case of multiple columns.

**Comment by BDonnot** [comment]
I have made some test to code an initial "grid2op backend" (first pre alpha version here https://github.com/BDonnot/pypowsybl/blob/grid2op_backend/pypowsybl/grid2op_backend/pypowsyblbackend.py ) and it does not seem to cause real issue.

To make sure it works as expected though, i need to call twice the `update_loads_with_data_frame` (once to change active value, once for the reactive) and twice the `update_generator_with_data_frame` (once for the active value, once for the voltage).

If it's easy to implement then for sure, being able to pass "Nan" for "I don't want to change" is a good thing. If it's too diffcult or too long to code, i"m not sure that's mandatory.

Good news is: this appear to be working as expected :-)

**Comment by sylvlecl** [comment]
Note that using `NaN` that way raises some questions:
1. it maybe be a valid value that we actually want to set
2. is it possible to use it for booleans ?
3. what would be the corresponding convention for strings ? (None, empty?)

To solve 1., we could add an option to the update function to select between the 2 behaviours (ignoring or setting the value), probably with default value = ignore.

Did not look into 2. and 3. :)

In any case once we are clear about those questions, it should be cheap to implement!

**Comment by BDonnot** [comment]
As far as I know:

1. If you get some "not a number (aka Nan)" for a load or generator, then the powerflow should return "not a number" for everything. That's a convention: when you add a number to Nan it results to Nan, same for multiplication, division, etc. So i'm not really sure it's relevant, from a "solver" point of view, to set something to "Nan". I may be wrong though.
2. At least in python, "Nan" is only available (as a specific value) for floating point. For integer, it should return the minimum integer, and for boolean it returns True (when i tested with numpy, se bellow). So it will not work for all data types, but only for floating point.
3. Good question, i believe `None` would be the right call there (because you can actually set something to be the empty string, see 1.)


Test for Nan for non floating points:
```python
import numpy as np
print(np.full(1, fill_value=np.NaN, dtype=bool))  # -> array([ True])
print(np.full(1, fill_value=np.NaN, dtype=int))  # -> array([-9223372036854775808])
print(np.full(1, fill_value=np.NaN, dtype=float))  # -> array([nan])
```

**Comment by sylvlecl** [comment]
Thanks for the test: it seems that we will have an issue for booleans and integers then.
A solution could be to provide a "mask" as a second argument but it becomes clumsy.

BUT, I just see that pandas has special `NA` values for [boolean](https://pandas.pydata.org/pandas-docs/stable/user_guide/boolean.html) and [int](https://pandas.pydata.org/pandas-docs/stable/user_guide/integer_na.html), maybe we could rely on it (need to check how it's implemented though, to be able to recognize those values on C++ side).

**Comment by AnkurArohi** [comment]
> @BDonnot right, we should support the nan to allow partial update in case of multiple columns.

I think this is still not implemented , right?
